### ISSUE TITLE
CGLIB lock contention/deadlock

### ISSUE THREAD

**Issue by felixbarny** [p6spy/p6spy#325]
After the deploy of a spring web application which includes the new stagemonitor 0.26.0 which depends on p6spy 2.3.0, the server did not respond after a short while.

The application was working fine with stagemonitor 0.25.0 which depends on p6spy 2.2.0. So the issue is either caused by a change in p6spy 2.3.0 or by the way stagemonitor uses p6spy in 0.26.0.

A thread dump revealed that tomcat threads are waiting for a CGLIB lock:
```
"ajp-bio-8031-exec-86" #237 daemon prio=5 os_prio=0 tid=0x00007f84dc037000 nid=0x1310 waiting for monitor entry [0x00007f84a31d6000]
   java.lang.Thread.State: BLOCKED (on object monitor)
        at com.p6spy.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:190)
        - waiting to lock <0x000000070e336be8> (a com.p6spy.cglib.core.AbstractClassGenerator$Source)
        at com.p6spy.cglib.proxy.Enhancer.createHelper(Enhancer.java:377)
        at com.p6spy.cglib.proxy.Enhancer.create(Enhancer.java:285)
        at com.p6spy.engine.proxy.ProxyFactory.createProxy(ProxyFactory.java:58)
        at com.p6spy.engine.logging.P6LogResultSetDelegate.invoke(P6LogResultSetDelegate.java:53)
        at com.p6spy.engine.proxy.GenericInvocationHandler.invoke(GenericInvocationHandler.java:116)
        at org.p6spy.$java.sql.PreparedStatement$$EnhancerByCGLIB$$3f082cc8.executeQuery(<generated>)
        at org.springframework.jdbc.core.JdbcTemplate$1.doInPreparedStatement(JdbcTemplate.java:688)
        at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:629)
        at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:680)
        at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:707)
        at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:757)
        at org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate.queryForObject(NamedParameterJdbcTemplate.java:211)
```

My suspicion is that this is because more proxies have to be generated now as `getResultSet` is now also instrumented on `Statement` and `PrepareStatement` see #320. 

When looking at `com.p6spy.cglib.core.AbstractClassGenerator#create`, you can see that there is a synchonized block inside the method:

```java
protected Object create(Object key) {
    [...]
    synchronized(this.source) {
        [...]
    }
    [...]
}
```

That means that the creation of all proxies is guarded by a global lock. A performance nightmare for multi threaded (web) applications when considering that a proxy will be created by p6spy each time a db call is performed.

I'd like to reintroduce the idea to not create proxy objects but instead write wrappers for the interfaces in `javax.sql.*`. This would be more verbose but also more performant and there won't be a chance of loader issues anymore. As this approach is not compatible with the current one (for example the `Connection`s returned by p6spy would not expose vendor specific methods) we could make this configureable. But maybe the proxy approach is too deeply integrated into p6spy so that it does not make sense to support both proxying and wrappers at the same time.

As this approach would take a while to implement and the issue is very urgent for me, do you have a idea for a quick fix? Maybe upgrading to CGLIB 3? Maybe cache the generated proxy class in `com.p6spy.engine.proxy.ProxyFactory#createProxy(T, com.p6spy.engine.proxy.GenericInvocationHandler<T>)` and not regenerate it again when the same `underlying` class is provided but only call `Class.newInstance`?

**Comment by felixbarny** [comment]
It seems like newer versions of CGLIB are caching generated classes automatically: https://github.com/cglib/cglib/commit/05f830fe682b0cfc18c5d30d9496e25e1bd1cac8

**Comment by typekpb** [comment]
@felixbarny the simplest way seems to be upgrading to latest cglib (although there seems to be some known issues: https://github.com/cglib/cglib/releases, anyway, it should be better than whatever we have). Feel free to give it a try.
Anyway, there is a test suite on p6spy side and you can obviously reproduce the problem in stagemonitor in rather short time.

btw: to me it sounds like stagemonitor is the coolest thing tha happened to p6spy lately :)

**Comment by typekpb** [comment]
fyi: for the p6spy part, I see no issues, tested `mvn clean test -DDB=H2` with:

    <dependency>
      <groupId>cglib</groupId>
      <artifactId>cglib-nodep</artifactId>
      <version>3.2.3</version>
    </dependency>

and all worked just fine

**Comment by felixbarny** [comment]
Which test suite do you mean? Could you reproduce the problem with cglib 2? Looking closer at the code it seems that generated classes are only reused when the Enhancer is reused. 

What do you think about the wrapper approach?

btw: thx ;)

**Comment by felixbarny** [comment]
Regarding the manual wrappers, see also https://github.com/p6spy/p6spy/issues/286#issuecomment-217025395.

**Comment by felixbarny** [comment]
Another problem with manual wrappers are the methods that where added in java 7. For example:

```
public class P6ConnectionWrapper implements Connection {
  public void setSchema(String schema) throws SQLException {
    delegate.setSchema(schema);
  }
}
```

P6Spy would have to be built with java 7 then. I'm not 100% sure but I think that the result should be java 6 compatible as long as you don't call `P6ConnectionWrapper#setSchema(String)` in Java 6 code. But as long as only the `Connection` interface is exposed, we should be goodl

**Comment by felixbarny** [comment]
I prototyped the wrapper approach which plays really nicely with the event architecture discussed in #319: https://github.com/felixbarny/p6spy/commit/db029899079f113bdf7a5a2fe11531332f032863

It looks like a lot of code, but most of it could be easily generated with IntelliJ and quickly edited with the multi cursor feature.

**Comment by quintonm** [comment]
I am not sure what could have changed in 2.3.0 that would have caused this issue.  I would look at commenting out the code that wraps the resultset to see if that solves the problem.  If not, then we might be looking in the wrong place.

**Comment by quintonm** [comment]
BTW - I can see concurrency being an issue for performance.  However, if a deadlock can occur, then I would think that it would be a defect either in the version of CGLIB that we are using or some problem in the way that we are using CGLIB.

**Comment by quintonm** [comment]
Fixed by #326
